tests/components/sharkiq/test_shark_iq.py

```diff
@@ -0,0 +1,283 @@
+
+try:
+    import ujson as json
+except ImportError:
+    import json
+
+
+MockAyla = MagicMock(spec=AylaApi)  # pylint: disable=invalid-name
+
+
+def _set_property(self, property_name, value):
+    """Set a property locally without hitting the API."""
+    if isinstance(property_name, enum.Enum):
+        property_name = property_name.value
+    if isinstance(value, enum.Enum):
+        value = value.value
+    self.properties_full[property_name]["value"] = value
+
+
+async def _async_set_property(self, property_name, value):
+    """Set a property locally without hitting the API."""
+    _set_property(self, property_name, value)
+
+
+def _get_mock_shark_vac(ayla_api: AylaApi) -> SharkIqVacuum:
+    """Create a crude sharkiq vacuum with mocked properties."""
+    shark = SharkIqVacuum(ayla_api, SHARK_DEVICE_DICT)
+    shark.properties_full = deepcopy(SHARK_PROPERTIES_DICT)
+    return shark
+
+
+async def _async_list_devices(_) -> List[Dict]:
+    """Generate a dummy of async_list_devices output."""
+    return [SHARK_DEVICE_DICT]
+
+
+@patch.object(SharkIqVacuum, "set_property_value", new=_set_property)
+@patch.object(SharkIqVacuum, "async_set_property_value", new=_async_set_property)
+async def test_shark_operation_modes(hass: HomeAssistant) -> None:
+    """Test all of the shark vacuum operation modes."""
+    ayla_api = MockAyla()
+    shark_vac = _get_mock_shark_vac(ayla_api)
+    coordinator = SharkIqUpdateCoordinator(hass, None, ayla_api, [shark_vac])
+    shark = SharkVacuumEntity(shark_vac, coordinator)
+
+    # These come from the setup
+    assert isinstance(shark.is_docked, bool) and not shark.is_docked
+    assert (
+        isinstance(shark.recharging_to_resume, bool) and not shark.recharging_to_resume
+    )
+    # Go through the operation modes while it's "off the dock"
+    await shark.async_start()
+    assert shark.operating_mode == shark.state == STATE_CLEANING
+    await shark.async_pause()
+    assert shark.operating_mode == shark.state == STATE_PAUSED
+    await shark.async_stop()
+    assert shark.operating_mode == shark.state == STATE_IDLE
+    await shark.async_return_to_base()
+    assert shark.operating_mode == shark.state == STATE_RETURNING
+
+    # Test the docked modes
+    await shark.async_stop()
+    shark.sharkiq.set_property_value(Properties.RECHARGING_TO_RESUME, 1)
+    shark.sharkiq.set_property_value(Properties.DOCKED_STATUS, 1)
+    assert isinstance(shark.is_docked, bool) and shark.is_docked
+    assert isinstance(shark.recharging_to_resume, bool) and shark.recharging_to_resume
+    assert shark.state == STATE_RECHARGING_TO_RESUME
+
+    shark.sharkiq.set_property_value(Properties.RECHARGING_TO_RESUME, 0)
+    assert shark.state == STATE_DOCKED
+
+    await shark.async_set_fan_speed("Eco")
+    assert shark.fan_speed == "Eco"
+    await shark.async_set_fan_speed("Max")
+    assert shark.fan_speed == "Max"
+    await shark.async_set_fan_speed("Normal")
+    assert shark.fan_speed == "Normal"
```

Comment:
Codacy found an issue: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/foreign-sub/home-assistant/pullRequest?prid=6090606)